File: leetcode/problems/cpp/minimum-absolute-difference-in-bst.cpp
// by @codeAbinash
// Time : O(n)
// Space : O(n)
#include "climits"
#include "algorithm"

struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right) : val(x), left(left), right(right) {}
};


class Solution {
  int ans = INT_MAX;
  TreeNode* prev = nullptr;
  void dfs(TreeNode* root) {
    if (!root) return;
    dfs(root->left);
    if (prev) ans = std::min(ans, root->val - prev->val);
    prev = root;
    dfs(root->right);
  }
public:
  int getMinimumDifference(TreeNode* root) {
    dfs(root);
    return ans;
  }
};

File: leetcode/problems/ts/minimum-absolute-difference-in-bst.ts
// by @codeAbinash
// Time : O(n)
// Space : O(n)

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function getMinimumDifference(root: TreeNode | null): number {
  const dfs = (node: TreeNode | null) => {
    if (!node) return;
    dfs(node.left);
    ans = Math.min(ans, Math.abs(node.val - prev));
    prev = node.val;
    dfs(node.right);
  };

  let ans = Number.MAX_VALUE;
  let prev = Number.MAX_VALUE;
  dfs(root);

  return ans;
}
